File: .skills/openclaw-skills/skills/purch-agent/agentic-commerce/scripts/shop.py
#!/usr/bin/env python3
"""
AI-powered shopping assistant using natural language.

Usage:
    python shop.py "<message>"

Example:
    python shop.py "comfortable running shoes under $100"
    python shop.py "wireless headphones with good noise cancellation"
"""

import argparse
import json
import re
import sys
import urllib.request
import urllib.error

BASE_URL = "https://api.purch.xyz"


def extract_variant_id(url: str) -> str | None:
    """Extract variant ID from Shopify product URL query parameter."""
    if not url:
        return None
    match = re.search(r'[?&]variant=(\d+)', url)
    return match.group(1) if match else None


def shop(message: str, context: dict | None = None) -> dict:
    """Use AI shopping assistant with natural language."""
    payload = {"message": message}
    if context:
        payload["context"] = context

    data = json.dumps(payload).encode("utf-8")

    try:
        req = urllib.request.Request(
            f"{BASE_URL}/shop",
            data=data,
            headers={"Content-Type": "application/json"},
            method="POST"
        )
        with urllib.request.urlopen(req, timeout=60) as response:
            return json.loads(response.read().decode())
    except urllib.error.HTTPError as e:
        error_body = e.read().decode() if e.fp else ""
        return {"error": f"HTTP {e.code}: {error_body}"}
    except Exception as e:
        return {"error": str(e)}


def main():
    parser = argparse.ArgumentParser(description="AI shopping assistant")
    parser.add_argument("message", help="Natural language shopping request")
    parser.add_argument("--price-min", type=float, help="Minimum price preference")
    parser.add_argument("--price-max", type=float, help="Maximum price preference")
    parser.add_argument("--json", action="store_true", help="Output raw JSON")

    args = parser.parse_args()

    # Build context if price preferences provided
    context = None
    if args.price_min is not None or args.price_max is not None:
        context = {"priceRange": {}}
        if args.price_min is not None:
            context["priceRange"]["min"] = args.price_min
        if args.price_max is not None:
            context["priceRange"]["max"] = args.price_max

    result = shop(args.message, context)

    if args.json:
        print(json.dumps(result, indent=2))
        return

    if "error" in result:
        print(f"❌ Error: {result['error']}")
        sys.exit(1)

    # Print AI reply
    reply = result.get("reply", "")
    if reply:
        print(f"🤖 {reply}")
        print()

    products = result.get("products", [])
    print(f"📦 Found {len(products)} products")
    print("-" * 60)

    for i, product in enumerate(products, 1):
        title = product.get("title", "Unknown")[:50]
        price = product.get("price", "N/A")
        currency = product.get("currency", "USD")
        rating = product.get("rating", "N/A")
        source = product.get("source", "unknown")
        url = product.get("productUrl", "")

        print(f"{i}. {title}")
        print(f"   💰 {currency} {price} | ⭐ {rating} | 🏷️ {source}")

        if source == "amazon":
            asin = product.get("asin", "N/A")
            print(f"   🛒 ASIN: {asin}")
        else:  # shopify
            variant_id = product.get("variantId") or extract_variant_id(url) or "N/A"
            clean_url = url.split('?')[0] if url else "N/A"
            print(f"   🛒 URL: {clean_url}")
            print(f"   🔖 Variant ID: {variant_id}")
        print()


if __name__ == "__main__":
    main()

File: .skills/openclaw-skills/skills/purch-agent/agentic-commerce/scripts/shop.ts
#!/usr/bin/env bun
/**
 * AI-powered shopping assistant using natural language.
 *
 * Usage:
 *   bun run shop.ts "<message>"
 *
 * Example:
 *   bun run shop.ts "comfortable running shoes under $100"
 *   bun run shop.ts "wireless headphones with good noise cancellation"
 */

const BASE_URL = "https://api.purch.xyz";

function extractVariantId(url: string | undefined): string | null {
  if (!url) return null;
  const match = url.match(/[?&]variant=(\d+)/);
  return match ? match[1] : null;
}

interface Product {
  asin: string;
  title: string;
  price: number;
  currency: string;
  rating?: number;
  reviewCount?: number;
  imageUrl?: string;
  productUrl?: string;
  source: string;
  vendor?: string;
  variantId?: string;
}

interface ShopResponse {
  reply: string;
  products: Product[];
}

interface ShopContext {
  priceRange?: { min?: number; max?: number };
  preferences?: string[];
}

async function shop(message: string, context?: ShopContext): Promise<ShopResponse> {
  const response = await fetch(`${BASE_URL}/shop`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ message, context }),
  });

  if (!response.ok) {
    const error = await response.text();
    throw new Error(`HTTP ${response.status}: ${error}`);
  }

  return response.json();
}

function parseArgs(args: string[]): {
  message: string;
  priceMin?: number;
  priceMax?: number;
  json?: boolean;
} {
  const result: ReturnType<typeof parseArgs> = { message: "" };
  let i = 0;

  while (i < args.length) {
    const arg = args[i];
    if (arg === "--price-min" && args[i + 1]) {
      result.priceMin = parseFloat(args[++i]);
    } else if (arg === "--price-max" && args[i + 1]) {
      result.priceMax = parseFloat(args[++i]);
    } else if (arg === "--json") {
      result.json = true;
    } else if (!arg.startsWith("--")) {
      result.message = arg;
    }
    i++;
  }

  return result;
}

async function main() {
  const args = process.argv.slice(2);

  if (args.length === 0 || args[0] === "--help") {
    console.log(`
Usage: bun run shop.ts "<message>" [options]

Options:
  --price-min <min>   Minimum price preference
  --price-max <max>   Maximum price preference
  --json              Output raw JSON

Example:
  bun run shop.ts "comfortable running shoes under $100"
    `);
    process.exit(0);
  }

  const options = parseArgs(args);

  if (!options.message) {
    console.error("❌ Error: Message is required");
    process.exit(1);
  }

  // Build context if price preferences provided
  let context: ShopContext | undefined;
  if (options.priceMin !== undefined || options.priceMax !== undefined) {
    context = { priceRange: {} };
    if (options.priceMin !== undefined) context.priceRange!.min = options.priceMin;
    if (options.priceMax !== undefined) context.priceRange!.max = options.priceMax;
  }

  try {
    const result = await shop(options.message, context);

    if (options.json) {
      console.log(JSON.stringify(result, null, 2));
      return;
    }

    if (result.reply) {
      console.log(`🤖 ${result.reply}`);
      console.log();
    }

    console.log(`📦 Found ${result.products.length} products`);
    console.log("-".repeat(60));

    result.products.forEach((product, i) => {
      const title = product.title.slice(0, 50);
      console.log(`${i + 1}. ${title}`);
      console.log(`   💰 ${product.currency} ${product.price} | ⭐ ${product.rating ?? "N/A"} | 🏷️ ${product.source}`);

      if (product.source === "amazon") {
        console.log(`   🛒 ASIN: ${product.asin}`);
      } else {
        // Shopify
        const variantId = product.variantId || extractVariantId(product.productUrl) || "N/A";
        const cleanUrl = product.productUrl?.split("?")[0] || "N/A";
        console.log(`   🛒 URL: ${cleanUrl}`);
        console.log(`   🔖 Variant ID: ${variantId}`);
      }
      console.log();
    });
  } catch (error) {
    console.error(`❌ Error: ${error}`);
    process.exit(1);
  }
}

main();
